File: src/Logic/Game.cpp
#include "Game.hpp"

namespace Game {
    Tank::Tank(TankID id, glm::vec2 pos) {
        this->id = id;
        this->pos = pos;
        
        size = tankSize;
        gunDirection = glm::vec2(0.0f, -1.0f);
        wheelDirection = glm::vec2(0.0f, -1.0f);
        speed = tankSpeed;

        isAlive = true;
    }

    Bullet::Bullet(BulletID id, glm::vec2 pos, glm::vec2 direction) {
        this->id = id;
        this->pos = pos;
        this->direction = direction;
    }

    BulletID Bullet::generateBulletID() {
        static BulletID id = 0;
        return id++;
    }

    World::World(State initialState) {
        state = initialState;
    }

    void World::tick(std::vector<Event> events) {
        state = getNextState(state, events);
    }

    State World::getState() const {
        return state;
    }

    bool intersects(Tank tank, Bullet bullet) {
        return tank.pos.x < bullet.pos.x && bullet.pos.x < tank.pos.x + tank.size.x &&
               tank.pos.y < bullet.pos.y && bullet.pos.y < tank.pos.y + tank.size.y;
    }

    bool bulletCollidesMapVertical(Bullet bullet) {
        return bullet.pos.y <= 0 || bullet.pos.y >= mapSize.y;
    }

    bool bulletCollidesMapHorizontal(Bullet bullet) {
        return bullet.pos.x <= 0 || bullet.pos.x >= mapSize.x;
    }

    State World::getNextState(State state, std::vector<Event> events) const {
        for(Event event : events) {
            auto stateTankIterator = state.tanks.find(event.tankId);

            if(stateTankIterator == state.tanks.end()) { continue; }

            Tank& tank = stateTankIterator->second;

            switch(event.type) {
                case EventType::FORWARD:
                    tank.pos += tank.wheelDirection * tank.speed;
                    tank.pos = glm::clamp(tank.pos, glm::vec2(0.0f, 0.0f), mapSize - tank.size);
                    break;
                case EventType::BACKWARD:
                    tank.pos -= tank.wheelDirection * tank.speed;
                    tank.pos = glm::clamp(tank.pos, glm::vec2(0.0f, 0.0f), mapSize - tank.size);
                    break;
                case EventType::SHOOT:
                    if(tank.bullets.size() >= maxBullets) { break; }
                    tank.bullets.push_back(Bullet(Bullet::generateBulletID(), tank.pos + tank.gunDirection * 2.f, tank.gunDirection));
                    break;
                case EventType::ROTATE_WHEELS_CLOCKWISE:
                    tank.wheelDirection = glm::rotate(tank.wheelDirection, -tankRotationSpeed);

                    // todo: remove this when gun is not tied to wheels
                    tank.gunDirection = tank.wheelDirection;
                    break;
                case EventType::ROTATE_WHEELS_CCLOCKWISE:
                    tank.wheelDirection = glm::rotate(tank.wheelDirection, tankRotationSpeed);

                    // todo: remove this when gun is not tied to wheels
                    tank.gunDirection = tank.wheelDirection;
                    break;
                case EventType::GUN_DIRECTION:
                    tank.gunDirection = glm::vec2(cos(event.value), sin(event.value));
                    break;
            }
        }

        std::vector<TankID> deadTanks;
        for(auto& tankPair : state.tanks) {
            Tank& tank = tankPair.second;

            for(std::size_t i = 0; i < tank.bullets.size(); i++) {
                Bullet& bullet = tank.bullets[i];
                bullet.pos += bullet.direction * bullet.speed;
                bullet.pos = glm::clamp(bullet.pos, glm::vec2(0.0f, 0.0f), mapSize);

                if(bulletCollidesMapVertical(bullet)) {
                    bullet.direction.y *= -1;
                    bullet.richochets++;
                }

                if(bulletCollidesMapHorizontal(bullet)) {
                    bullet.direction.x *= -1;
                    bullet.richochets++;
                }

                if(bullet.richochets > bullet.maxRichochets) {
                    tank.bullets.erase(tank.bullets.begin() + i--);
                    continue;
                }

                for(auto& otherTankPair : state.tanks) {
                    TankID otherTankId = otherTankPair.first;
                    Tank& otherTank = otherTankPair.second;

                    if(intersects(otherTank, bullet)) {
                        tank.bullets.erase(tank.bullets.begin() + i--);
                        deadTanks.push_back(otherTankId);
                        break;
                    }
                }
            }
        }

        return state;
    }
}

File: src/game.rs
extern crate nalgebra_glm as glm;

use glm::{pi, rotate_vec2, Vec2, Vec3};
use std::collections::HashMap;
use std::sync::atomic::{AtomicUsize, Ordering};

pub type TankID = i32;
pub type BulletID = i32;

pub const GAME_FREQUENCY: f32 = 300.0;
pub fn tank_size() -> Vec2 {
    Vec2::new(10.0, 10.0)
}

pub const TANK_SPEED: f32 = 200.0 / GAME_FREQUENCY / 4.0;
pub const BULLET_SPEED: f32 = 1.5 * TANK_SPEED;
pub fn tank_rotation_speed() -> f32 {
    2.0 * pi::<f32>() / GAME_FREQUENCY / 1.5
}
pub const MAX_BULLETS: u16 = 5;
pub fn map_size() -> Vec2 {
    Vec2::new(200.0, 100.0)
}

#[derive(Clone)]
pub struct Bullet {
    pub id: BulletID,
    pub pos: Vec2,
    pub direction: Vec2,
    pub speed: f32,
    pub ricochets: u16,
    pub max_ricochets: u16,
}

impl Bullet {
    pub fn new(id: BulletID, pos: Vec2, direction: Vec2) -> Self {
        Self {
            id,
            pos,
            direction,
            speed: BULLET_SPEED,
            ricochets: 0,
            max_ricochets: 1,
        }
    }

    pub fn generate_bullet_id() -> BulletID {
        static ID: AtomicUsize = AtomicUsize::new(0);
        ID.fetch_add(1, Ordering::SeqCst) as BulletID
    }
}

#[derive(Clone)]
pub struct Tank {
    pub id: TankID,
    pub pos: Vec2,
    pub size: Vec2,
    pub gun_direction: Vec2,
    pub wheel_direction: Vec2,
    pub speed: f32,
    pub is_alive: bool,
    pub bullets: Vec<Bullet>,
}

impl Tank {
    pub fn new(id: TankID, pos: Vec2) -> Self {
        Self {
            id,
            pos,
            size: tank_size(),
            gun_direction: Vec2::new(0.0, -1.0),
            wheel_direction: Vec2::new(0.0, -1.0),
            speed: TANK_SPEED,
            is_alive: true,
            bullets: Vec::new(),
        }
    }
}

#[derive(Clone)]
pub struct State {
    pub tanks: HashMap<TankID, Tank>,
}

#[derive(Copy, Clone, Hash, PartialEq)]
pub enum EventType {
    Forward,
    Backward,
    Shoot,
    RotateWheelsClockwise,
    RotateWheelsCClockwise,
    // GunDirection,
}

#[derive(Clone)]
pub struct Event {
    pub tank_id: i32,
    pub event_type: EventType,

    pub value: f32,
}

pub struct World {
    state: State,
}

impl World {
    pub fn new(initial_state: State) -> Self {
        Self {
            state: initial_state,
        }
    }

    pub fn tick(&mut self, events: Vec<Event>) {
        self.state = self.get_next_state(self.state.clone(), events);
    }

    pub fn get_state(&self) -> &State {
        &self.state
    }
}

impl World {
    fn bullet_collides_map_vertical(bullet: &Bullet) -> bool {
        bullet.pos.y <= 0.0 || bullet.pos.y >= map_size().y
    }

    fn bullet_collides_map_horizontal(bullet: &Bullet) -> bool {
        bullet.pos.x <= 0.0 || bullet.pos.x >= map_size().x
    }

    fn intersects(tank: &Tank, bullet: &Bullet) -> bool {
        tank.pos.x < bullet.pos.x
            && bullet.pos.x < tank.pos.x + tank.size.x
            && tank.pos.y < bullet.pos.y
            && bullet.pos.y < tank.pos.y + tank.size.y
    }

    pub fn clamp3(v: Vec3, min: Vec3, max: Vec3) -> Vec3 {
        Vec3::new(
            v.x.max(min.x).min(max.x),
            v.y.max(min.y).min(max.y),
            v.z.max(min.z).min(max.z),
        )
    }

    pub fn clamp2(v: Vec2, min: Vec2, max: Vec2) -> Vec2 {
        Vec2::new(v.x.max(min.x).min(max.x), v.y.max(min.y).min(max.y))
    }

    pub fn get_next_state(&self, mut state: State, events: Vec<Event>) -> State {
        for event in events {
            if let Some(tank) = state.tanks.get_mut(&event.tank_id) {
                match event.event_type {
                    EventType::Forward => {
                        tank.pos += tank.wheel_direction * tank.speed;
                        tank.pos =
                            Self::clamp2(tank.pos, Vec2::new(0.0, 0.0), map_size() - tank.size);
                    }
                    EventType::Backward => {
                        tank.pos -= tank.wheel_direction * tank.speed;
                        tank.pos =
                            Self::clamp2(tank.pos, Vec2::new(0.0, 0.0), map_size() - tank.size);
                    }
                    EventType::Shoot => {
                        if tank.bullets.len() < MAX_BULLETS as usize {
                            tank.bullets.push(Bullet::new(
                                Bullet::generate_bullet_id(),
                                tank.pos + tank.gun_direction * 2.0,
                                tank.gun_direction,
                            ));
                        }
                    }
                    EventType::RotateWheelsClockwise => {
                        tank.wheel_direction =
                            rotate_vec2(&tank.wheel_direction, -tank_rotation_speed());
                        tank.gun_direction = tank.wheel_direction;
                    }
                    EventType::RotateWheelsCClockwise => {
                        tank.wheel_direction =
                            rotate_vec2(&tank.wheel_direction, tank_rotation_speed());
                        tank.gun_direction = tank.wheel_direction;
                    } 
                    /* EventType::GunDirection => {
                        tank.gun_direction = Vec2::new(event.value.cos(), event.value.sin());
                    } */
                }
            }
        }

        for (_, tank) in &mut state.tanks {
            for i in 0..tank.bullets.len() {
                let bullet = &mut tank.bullets[i];
                bullet.pos += bullet.direction * bullet.speed;
                bullet.pos = Self::clamp2(bullet.pos, Vec2::new(0.0, 0.0), map_size());

                if Self::bullet_collides_map_vertical(bullet) {
                    bullet.direction.y *= -1.0;
                    bullet.ricochets += 1;
                }

                if Self::bullet_collides_map_horizontal(bullet) {
                    bullet.direction.x *= -1.0;
                    bullet.ricochets += 1;
                }
            }
        }

        let mut dead_tanks = Vec::new();
        let mut dead_bullets: Vec<(i32, usize)> = Vec::new();
        for (tank_id, tank) in &state.tanks {
            for i in (0..tank.bullets.len()).rev() {
                let bullet = &tank.bullets[i];

                if bullet.ricochets > bullet.max_ricochets {
                    dead_bullets.push((*tank_id, i));
                    continue;
                }

                for (other_tank_id, other_tank) in &state.tanks {
                    if Self::intersects(&other_tank, bullet) {
                        dead_bullets.push((*tank_id, i));
                        dead_tanks.push(*other_tank_id);
                        break;
                    }
                }
            }
        }

        for (tank_id, bullet_id) in dead_bullets {
            state
                .tanks
                .get_mut(&tank_id)
                .unwrap()
                .bullets
                .remove(bullet_id);
        }

        for dead_tank in dead_tanks {
            state.tanks.remove(&dead_tank);
        }

        state
    }
}
